src/sdam/server.ts

```diff
@@ -368,9 +370,10 @@ export class Server extends TypedEventEmitter<ServerEvents> {
       if (session?.pinnedConnection !== conn) {
         if (reauthPromise != null) {
           // The reauth promise only exists if it hasn't thrown.
-          void reauthPromise.finally(() => {
+          const checkBackIn = () => {
             this.pool.checkIn(conn);
-          });
+          };
+          void reauthPromise.then(checkBackIn, checkBackIn);
```

Comment:
`.finally` just responds to either outcome and the rejection continues to be unhandled

Reply:
It seems wrong that we're manually attaching listeners to a promise instead of a try-catch block, and doubly wrong that we're adding the finally listener so far from the origin of the promise.  Should we instead be adding this logic in the `finally` before we await the reauth promise (maybe also putting this logic in an async function?)?

```typescript

async function reauthenticate(connection) {
  try {
    await pool.reauthenticate(connection);
  } finally {
    pool.checkIn(connection);
  }
}

// server.command
await abortable(reauthenticate(connection), options);
```